src/anndata/_io/specs/methods.py

```diff
@@ -1028,30 +1034,82 @@ def write_nullable_integer(
     _writer.write_elem(g, "values", v._data, dataset_kwargs=dataset_kwargs)
 
 
+@_REGISTRY.register_write(
+    H5Group, pd.arrays.StringArray, IOSpec("nullable-string-array", "0.1.0")
+)
+@_REGISTRY.register_write(
+    ZarrGroup, pd.arrays.StringArray, IOSpec("nullable-string-array", "0.1.0")
+)
+def write_nullable_string(
+    f: GroupStorageType,
+    k: str,
+    v: pd.arrays.StringArray,
+    *,
+    _writer: Writer,
+    dataset_kwargs: Mapping[str, Any] = MappingProxyType({}),
+):
+    from ..._settings import settings
+
+    if not settings.allow_write_nullable_strings:
+        msg = (
+            "`anndata.settings.allow_write_nullable_strings` is False, "
+            "because writing of `pd.arrays.StringArray` is new "
+            "and not supported in anndata < 0.11, still use by many people. "
+            "Opt-in to writing these arrays by toggling the setting to True."
+        )
+        raise RuntimeError(msg)
+    g = f.require_group(k)
+    if (mask := v.isna()).any():
+        _writer.write_elem(g, "mask", mask, dataset_kwargs=dataset_kwargs)
+    _writer.write_elem(
+        g, "values", v.fillna("")._ndarray, dataset_kwargs=dataset_kwargs
+    )
+
+
 @_REGISTRY.register_read(H5Group, IOSpec("nullable-integer", "0.1.0"))
 @_REGISTRY.register_read(ZarrGroup, IOSpec("nullable-integer", "0.1.0"))
 def read_nullable_integer(
     elem: GroupStorageType, *, _reader: Reader
 ) -> pd.api.extensions.ExtensionArray:
-    if "mask" in elem:
-        return pd.arrays.IntegerArray(
-            _reader.read_elem(elem["values"]), mask=_reader.read_elem(elem["mask"])
-        )
-    else:
-        return pd.array(_reader.read_elem(elem["values"]))
+    return _read_nullable(elem, _reader=_reader, array_type=pd.arrays.IntegerArray)
 
 
 @_REGISTRY.register_read(H5Group, IOSpec("nullable-boolean", "0.1.0"))
 @_REGISTRY.register_read(ZarrGroup, IOSpec("nullable-boolean", "0.1.0"))
 def read_nullable_boolean(
     elem: GroupStorageType, *, _reader: Reader
 ) -> pd.api.extensions.ExtensionArray:
-    if "mask" in elem:
-        return pd.arrays.BooleanArray(
-            _reader.read_elem(elem["values"]), mask=_reader.read_elem(elem["mask"])
-        )
-    else:
-        return pd.array(_reader.read_elem(elem["values"]))
+    return _read_nullable(elem, _reader=_reader, array_type=pd.arrays.BooleanArray)
```

Comment:
I think we have semantics for this sort of repeated implementation

https://github.com/scverse/anndata/blob/main/src/anndata/_io/specs/methods.py#L1120-L1121

So maybe just adding a `partial` here to specify the `array_type` would do it?

Reply:
this would prevent us from re-implementing the same function signature a bunch